base/leaky_bucket.go

```diff
@@ -386,6 +390,74 @@ func (b *LeakyBucket) IsSupported(feature sgbucket.BucketFeature) bool {
 	return b.bucket.IsSupported(feature)
 }
 
+func (b *LeakyBucket) Query(statement string, params interface{}, consistency gocb.ConsistencyMode, adhoc bool) (results gocb.QueryResults, err error) {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return nil, errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.Query(statement, params, consistency, adhoc)
+}
+
+func (b *LeakyBucket) ExplainQuery(statement string, params interface{}) (plain map[string]interface{}, err error) {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return nil, errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.ExplainQuery(statement, params)
+}
+
+func (b *LeakyBucket) CreateIndex(indexName string, expression string, filterExpression string, options *N1qlIndexOptions) error {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.CreateIndex(indexName, expression, filterExpression, options)
+}
+
+func (b *LeakyBucket) BuildDeferredIndexes(indexSet []string) error {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.BuildDeferredIndexes(indexSet)
+}
+
+func (b *LeakyBucket) CreatePrimaryIndex(indexName string, options *N1qlIndexOptions) error {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.CreatePrimaryIndex(indexName, options)
+}
+
+func (b *LeakyBucket) WaitForIndexOnline(indexName string) error {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.WaitForIndexOnline(indexName)
+}
+
+func (b *LeakyBucket) GetIndexMeta(indexName string) (exists bool, meta *gocb.IndexInfo, err error) {
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
+	if !ok {
+		return false, nil, errors.New("Not GOCB Bucket")
+	}
+	return gocbBucket.GetIndexMeta(indexName)
+}
+
+func (b *LeakyBucket) DropIndex(indexName string) error {
+	if b.config.DropIndexErrorCount != 0 {
+		b.config.DropIndexErrorCount--
+		return errors.New(fmt.Sprintf("Artificial leaky bucket error %d fails remaining", b.config.DropIndexErrorCount))
+	}
+	gocbBucket, ok := AsGoCBBucket(b.bucket)
```

Comment:
Perhaps do the bucket type check first, before we do the artificial error stuff.